e2e_tests/tests/experiment/test_custom_searcher.py

```diff
@@ -219,3 +226,305 @@ def sample_params(self) -> Dict[str, int]:
+                self.closed_trials.add(request_id)
+        else:
+            next_rung = self.rungs[rung_idx + 1]
+            logging.info(f"Promoting in rung {rung_idx}")
+            for promoted_request_id in rung.promotions_async(request_id, metric, self.divisor):
+                self.trial_rungs[promoted_request_id] = rung_idx + 1
+                next_rung.outstanding_trials += 1
+                if promoted_request_id not in self.early_exit_trials:
+                    logging.info(f"Promoted {promoted_request_id}")
+                    units_needed = max(next_rung.units_needed - rung.units_needed, 1)
+                    ops.append(ValidateAfter(promoted_request_id, units_needed))
+                    added_train_workload = True
+                    self.pending_trials += 1
+                else:
+                    return self.promote_async(promoted_request_id, sys.float_info.max)
+
+        all_trials = len(self.trial_rungs) - self.invalid_trials
+        if not added_train_workload and all_trials < self.max_trials:
+            logging.info("Creating new trial instead of promoting")
+            self.pending_trials += 1
+
+            create = Create(
+                request_id=uuid.uuid4(),
+                hparams=self.sample_params(),
+                checkpoint=None,
+            )
+            ops.append(create)
+            ops.append(
+                ValidateAfter(request_id=create.request_id, length=self.rungs[0].units_needed)
+            )
+
+            self.trial_rungs[create.request_id] = 0
+
+        if len(self.rungs[0].metrics) == self.max_trials:
+            ops.extend(self._get_close_rungs_ops())
+
+        return ops
+
+    def _get_close_rungs_ops(self) -> List[Operation]:
+        ops: List[Operation] = []
+
+        for rung in self.rungs:
+            if rung.outstanding_trials > 0:
+                break
+            for trial_metric in rung.metrics:
+                if not trial_metric.promoted and trial_metric.request_id not in self.closed_trials:
+                    if trial_metric.request_id not in self.early_exit_trials:
+                        logging.info(f"Closing trial {trial_metric.request_id}")
+                        ops.append(Close(trial_metric.request_id))
+                        self.closed_trials.add(trial_metric.request_id)
+        return ops
+
+    def sample_params(self) -> Dict[str, object]:
+        hparams = {
+            "global_batch_size": 10,
+            "metrics_base": np.random.uniform(0.2, 0.7),
+            "metrics_progression": "decreasing",
+        }
+        logging.info(f"hparams={hparams}")
+        return hparams
+
+    def progress(self) -> float:
+        if 0 < self.max_concurrent_trials < self.pending_trials:
+            raise RuntimeError("Pending trial is greater than max concurrent trials")
+
+        all_trials = len(self.rungs[0].metrics)
+        progress = all_trials / (1.2 * self.max_trials)
```

Comment:
this was intended to be a lower bound estimate on progress and there may be better alternatives. let me know if you have suggestions.